agent/consul/multilimiter/multilimiter_test.go

```diff
@@ -30,58 +31,168 @@ func TestNewMultiLimiter(t *testing.T) {
 func TestRateLimiterUpdate(t *testing.T) {
-	c := Config{LimiterConfig: LimiterConfig{Rate: 0.1}, ReconcileCheckLimit: 1 * time.Millisecond, ReconcileCheckInterval: 10 * time.Millisecond}
+	c := Config{LimiterConfig: LimiterConfig{Rate: 0.1}, ReconcileCheckLimit: 1 * time.Hour, ReconcileCheckInterval: 10 * time.Millisecond}
 	m := NewMultiLimiter(c)
+	m.Run(context.Background())
 	key := makeKey([]byte("test"))
+
+	//Allow a key
 	m.Allow(Limited{key: key})
+	time.Sleep(100 * time.Millisecond)
 	limiters := m.limiters.Load()
 	l1, ok1 := limiters.Get(key)
+
+	// check key exist
 	require.True(t, ok1)
 	require.NotNil(t, l1)
+
 	la1 := l1.(*Limiter).lastAccess.Load()
+
+	// allow same key again
+	time.Sleep(10 * time.Millisecond)
 	m.Allow(Limited{key: key})
 	limiters = m.limiters.Load()
 	l2, ok2 := limiters.Get(key)
+
+	// check it exist and it's same key
 	require.True(t, ok2)
 	require.NotNil(t, l2)
 	require.Equal(t, l1, l2)
+
+	// last access should be different
 	la2 := l1.(*Limiter).lastAccess.Load()
-	require.Equal(t, la1, la2)
+	require.NotEqual(t, la1, la2)
 
 }
 
 func TestRateLimiterCleanup(t *testing.T) {
 
 	// Create a limiter and Allow a key, check the key exists
-	c := Config{LimiterConfig: LimiterConfig{Rate: 0.1}, ReconcileCheckLimit: 1 * time.Millisecond, ReconcileCheckInterval: 10 * time.Millisecond}
+	c := Config{LimiterConfig: LimiterConfig{Rate: 0.1}, ReconcileCheckLimit: 1 * time.Second, ReconcileCheckInterval: 10 * time.Millisecond}
 	m := NewMultiLimiter(c)
+	limiters := m.limiters.Load()
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()
 	m.Run(ctx)
 	key := makeKey([]byte("test"))
 	m.Allow(Limited{key: key})
-	limiters := m.limiters.Load()
+	retry.Run(t, func(r *retry.R) {
+		time.Sleep(100 * time.Millisecond)
+		l := m.limiters.Load()
+		require.NotEqual(r, limiters, l)
+		limiters = l
+	})
+
 	l, ok := limiters.Get(key)
 	require.True(t, ok)
 	require.NotNil(t, l)
 
 	// Wait > ReconcileCheckInterval and check that the key was cleaned up
-	time.Sleep(20 * time.Millisecond)
-	limiters = m.limiters.Load()
+	retry.Run(t, func(r *retry.R) {
+		time.Sleep(100 * time.Millisecond)
+		l := m.limiters.Load()
+		require.NotEqual(r, limiters, l)
+		limiters = l
+	})
 	l, ok = limiters.Get(key)
 	require.False(t, ok)
 	require.Nil(t, l)
 
 	// Stop the cleanup routine, check that a key is not cleaned up after > ReconcileCheckInterval
 	cancel()
+
 	m.Allow(Limited{key: key})
-	time.Sleep(20 * time.Millisecond)
-	limiters = m.limiters.Load()
+	txn := m.limiters.Load().Txn()
+	m.runStoreOnce(context.Background(), time.NewTicker(1*time.Microsecond), txn)
+	m.runStoreOnce(context.Background(), time.NewTicker(1*time.Microsecond), txn)
+	retry.Run(t, func(r *retry.R) {
+		time.Sleep(100 * time.Millisecond)
+		l := m.limiters.Load()
+		require.NotEqual(r, limiters, l)
+		limiters = l
+	})
 	l, ok = limiters.Get(key)
 	require.True(t, ok)
 	require.NotNil(t, l)
 }
 
+func TestRateLimiterStore(t *testing.T) {
```

Comment:
Totally optional/take with a pinch of salt 🧂 

I wonder if it'd be helpful to make the `Ticker` an interface that can be stubbed out in tests. That way you could call the (exported) `Run` method and have the tests trigger "ticks" rather than calling the unexported method directly. It would also mean you wouldn't need to sleep in tests (assuming you had some kind of commit callback as in my previous comment).

Reply:
I tried this and I was partially successful. I was not able to create a direct interface to Ticker because it expose a channel directly so I needed to create a wrapper around it. I also don't like to expose that wrapper outside the package as its only purpose is testing. So I ended up with an hybrid solution where I can mock the ticker in some tests but need to use a real ticker in others, which needed to test the `Run(ctx)` API call. For those I used a retry instead of a sleep.